test/ngRoute/routeSpec.js

```diff
@@ -2082,4 +2082,58 @@ describe('$route', function() {
       expect(function() { $route.updateParams(); }).toThrowMinErr('ngRoute', 'norout');
     }));
   });
+
+  describe('testability', function() {
+    it('should wait for route promises before calling callbacks', function() {
+      var deferred;
+
+      module(function($provide, $routeProvider) {
+        $routeProvider.when('/path', { template: '', resolve: {
+          a: function($q) {
+            deferred = $q.defer();
+            return deferred.promise;
+          }
+        } });
+      });
+
+      inject(function($location, $route, $rootScope, $httpBackend, $$testability) {
+        $location.path('/path');
+        $rootScope.$digest();
+
+        var callback = jasmine.createSpy('callback');
+        $$testability.whenStable(callback);
+        expect(callback).not.toHaveBeenCalled();
+
+        deferred.resolve();
+        $rootScope.$digest();
+        expect(callback).toHaveBeenCalled();
+      });
+    });
+
+    it('should call callback after route promises are rejected', function() {
```

Comment:
@gkalpak I wrote this unit test for the race condition I was worried about, and it does appear to be a problem